src/msw/printdlg.cpp

```diff
@@ -689,6 +768,16 @@ bool wxWindowsPrintNativeData::TransferFrom( const wxPrintData &data )
     return true;
 }
 
+wxPrintData GetDefaultPrintData(const wxString& printerName = wxEmptyString) {
```

Comment:
This function currently wouldn't be usable even in DLL builds under MSW, so it's really not very useful. We could define it in `wxWindowsPrintDialog` (as a static method) instead.

Reply:
hmm, do we have a place for all ports so that other ports can implement that method too? Currently it would be ok to simply return wxPrintData() on other ports because this is the current default. Maybe in wxPrintFactory?

Reply:
In my latest commit I've moved the function to the wxWindowsPrintDialog class and the function is static, but I would really like to see this method in al ports.

Reply:
I'm not sure what would it do in all ports, at least under Linux I don't think the previously stored values are saved anywhere unless the application does it itself. I have no idea under macOS though, does anybody know if the last used values are preserved automatically there?

Ideal would be to have some higher level API that could be used to save/restore print settings if necessary (see also [this ticket](https://trac.wxwidgets.org/ticket/19178)), allowing to write the code which works on all platforms optimally, i.e. avoiding saving `wxPrintData` in `wxConfig` under MSW (where it's unnecessary, if I understand correctly), but doing it elsewhere. But I don't see how would it work right now, so perhaps we could apply the rest of the changes here for now and leave the rest for another PR?

Reply:
>  so perhaps we could apply the rest of the changes here for now and leave the rest for another PR?
Sure. This is a good way to proceed.

> does anybody know if the last used values are preserved automatically
Not automatically, manually. Users and admins can set preferences manually, so that when you open a print dialog next time, the dialog is preset according to these preferences.
![image](https://user-images.githubusercontent.com/33447587/125208835-2a24ee80-e295-11eb-9537-64995f9e3dfe.png)

These preferences are also very helpful when you e.g. use the "print" verb in ShellExecute to print a file on a printer. Then application can then load these preferences and the file is printed according to these preferences and so you do not need a bunch of command line arguments to tell the application how to print the file.

Reply:
OK, so, just to confirm, this doesn't free the program from the need to save/restore the settings used in the print dialogs by the user?

If so, how exactly is this supposed to work with the default settings: suppose the user has changed the page orientation to landscape. If the program saves the full `wxPrintData` to config, restoring it later would overwrite any defaults. So I guess we need to save only the actually changed fields, then get the default print data first and restore these fields only?

Reply:
> OK, so, just to confirm, this doesn't free the program from the need to save/restore the settings used in the print dialogs by the user?

The app can continue to do this if it makes sense.

> So I guess we need to save only the actually changed fields, then get the default print data first and restore these fields only?

It could be done like that, but I think it complicates things. I think a hierarchical approach is sufficient. In the end there can be several storage locations for a configuration. It currently looks like this under Windows:

1. The app itself can store the print configuration if the user wants that and if the app supports that maybe via a checkbox. These settings are _app specific_.
2. The user can set printing preferences for a printer stored in the HKCU area of the Windows registry. These preferences would then be _available to all apps_.
3. Admins can set printing preferences for a printer stored in the HKLM area of the Windows registry. These preferences would then be _available to all apps and all users_.
4. The printer driver itself also has default settings for the printer and _these will be used if no other settings are saved in any of the above locations_.

This is a hierarchy of locations and the app can look down from the top and take the settings that come first.

Reply:
> So I guess we need to save only the actually changed fields, then get the default print data first and restore these fields only?

That could at least work for known/common fields, but not the vendor-specific binary blob which can contain influental settings such as "enable borderless printing". Even if an application tries its best (whatever that is, there may be guidelines) as a user I will still check all setings prior to print preview and again before printing.